src/geom/prep/PreparedPolygonPredicate.cpp

```diff
@@ -37,71 +38,118 @@ namespace prep { // geos.geom.prep
 //
 // protected:
 //
-bool
-PreparedPolygonPredicate::isAllTestComponentsInTarget(const geom::Geometry* testGeom) const
-{
-    geom::Coordinate::ConstVect pts;
-    geom::util::ComponentCoordinateExtracter::getCoordinates(*testGeom, pts);
-
-    for(std::size_t i = 0, ni = pts.size(); i < ni; i++) {
-        const geom::Coordinate* pt = pts[i];
-        const int loc = prepPoly->getPointLocator()->locate(pt);
-        if(geom::Location::EXTERIOR == loc) {
-            return false;
+struct LocationMatchingFilter : public GeometryComponentFilter {
+    explicit LocationMatchingFilter(algorithm::locate::PointOnGeometryLocator* locator, int loc) :
+        pt_locator(locator), test_loc(loc), found(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    const int test_loc;
+    bool found;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        const int loc = pt_locator->locate(pt);
+
+        if (loc == test_loc) {
+            found = true;
+        }
+    }
+
+    bool isDone() override {
+        return found;
+    }
+};
+
+struct LocationNotMatchingFilter : public GeometryComponentFilter {
+    explicit LocationNotMatchingFilter(algorithm::locate::PointOnGeometryLocator* locator, int loc) :
+            pt_locator(locator), test_loc(loc), found(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    const int test_loc;
+    bool found;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        const int loc = pt_locator->locate(pt);
+
+        if (loc != test_loc) {
+            found = true;
+        }
+    }
+
+    bool isDone() override {
+        return found;
+    }
+};
+
+struct OutermostLocationFilter : public GeometryComponentFilter {
+    explicit OutermostLocationFilter(algorithm::locate::PointOnGeometryLocator* locator) :
+    pt_locator(locator),
+    outermost_loc(geom::Location::UNDEF),
+    done(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    Location::Value outermost_loc;
+    bool done;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        auto loc = static_cast<Location::Value>(pt_locator->locate(pt));
+
+        if (outermost_loc == Location::UNDEF || outermost_loc == Location::INTERIOR) {
+            outermost_loc = loc;
+        } else if (Location::EXTERIOR) {
```

Comment:
As @strk pointed out, this looks like a bug.  Shouldn't there be a mention of BOUNDARY here?

Would be good to find a test case that fails because of this.

And it would probably be nice to name this logic in a separate method (something like `locMax`).  Or move it into the Location class.

Reply:
I do see the bug @strk reported, but I don't think BOUNDARY needs to be mentioned. I'll try to find a test case.

Reply:
Fixed in 67d8a27 with test case. Would be fine to move in to `Location` but I would probably then revert the implementation back to `std::max`.

Reply:
Great. Since the code doesn't make it fully clear that there is an ordering relation on Location, it would be better to define this explicitly in the Location class, and then use that here. Perhaps a method Location::Index? That would allow a simple comparison to be used in this method.

It would also be nice to run all the predicate unit tests through the Prepared code. Not sure if there's a portable way to do this. A quick-n-dirty way is just to hack the XMLTester code for each predicate to run both.

One reason to do this is to validate that the predicate unit tests really have full coverage. If they don't trigger this bug that's an issue that needs to be fixed.